### PR TITLE
Store column values of not cache-able associations

### PR DESCRIPTION
**Pull request by goetas**
Fixes #1301

### CODE REVIEW

File: lib/Doctrine/ORM/Cache/DefaultEntityHydrator.php [hunk 1]
```diff
@@ -73,8 +73,15 @@ public function buildCacheEntry(ClassMetadata $metadata, EntityCacheKey $key, $e
             }
 
             if ( ! isset($assoc['cache']) || ! ($assoc['type'] & ClassMetadata::TO_ONE)) {
+                $associatedEntity = $data[$name];
                 unset($data[$name]);
 
+                if ($this->uow->isInIdentityMap($associatedEntity)) {
```
Comment (goetas) [inline lib/Doctrine/ORM/Cache/DefaultEntityHydrator.php]: What if the entity is not in identity map?

Comment (goetas) [inline lib/Doctrine/ORM/Cache/DefaultEntityHydrator.php]: has it to be persisted somehow?

Comment (bakura10) [inline lib/Doctrine/ORM/Cache/DefaultEntityHydrator.php]: I'm not an expert but I think some kind of proxy should be created to allow traversal of the association (that will obvisouly trigger another request).

Comment (goetas) [inline lib/Doctrine/ORM/Cache/DefaultEntityHydrator.php]: you are right! `IdentifierFlattener` is the answer! I will try do do it

File: tests/Doctrine/Tests/ORM/Functional/SecondLevelCacheOneToOneTest.php [hunk 2]
```diff
@@ -187,4 +189,32 @@ public function testPutAndLoadOneToOneBidirectionalRelation()
...
+        $this->_em->persist($token);
+        $this->_em->flush();
+        $this->_em->clear();
+
+        $queryCount = $this->getCurrentQueryCount();
+
+        $this->assertTrue($this->cache->containsEntity(Token::CLASSNAME, $token->getToken()));
+        $this->assertFalse($this->cache->containsEntity(Client::CLASSNAME, $client->getId()));
+
+        $entity = $this->_em->find(Token::CLASSNAME, $token->getToken());
```
Comment (bakura10) [inline tests/Doctrine/Tests/ORM/Functional/SecondLevelCacheOneToOneTest.php]: Can you try to remove this line and make sure the test still passes? From your code, it seems that you are trying to extract the identifier from the association, but if the association is not cached, then this means the cache does not hold any identifier for the association, right?

**Comment by bakura10** [comment]
It would be nice if you could also add the same kind of tests but for ManyToOne associations, to make sure it also work properly when there is no @Cache on the association.

**Comment by goetas** [comment]
Done

**Comment by guilhermeblanco** [comment]
@goetas at first this seems very good. I may have looked too quickly, but do you mind ensuring there's a & TO_ONE association with composite keys and check if it works as a test case?

**Comment by goetas** [comment]
I think that if I add a "good" association with multiple keys and nested association, the test will fail.

Any chance to see #1113 merged before this? The mentioned PR contains a lot of improvements for composite keys that will be helpful to create a good test for this PR.

**Comment by bakura10** [comment]
@goetas did you try to run the test by commenting the line I've told in the comment?

**Comment by goetas** [comment]
@bakura10 you mean this right?

```php
$this->assertNull($this->cache->getEntityCacheRegion(Client::CLASSNAME));
$this->assertInstanceOf('Doctrine\ORM\Cache\Region', $this->cache->getEntityCacheRegion(Token::CLASSNAME));

$client = new Client('FabioBatSilva');
$token  = new Token('token-hash', $client);

$this->_em->persist($client);
$this->_em->persist($token);
$this->_em->flush();
//$this->_em->clear();

$queryCount = $this->getCurrentQueryCount();

$this->assertTrue($this->cache->containsEntity(Token::CLASSNAME, $token->getToken()));
$this->assertFalse($this->cache->containsEntity(Client::CLASSNAME, $client->getId()));

$entity = $token;

$this->assertInstanceOf(Token::CLASSNAME, $entity);
$this->assertInstanceOf(Client::CLASSNAME, $entity->getClient());
$this->assertEquals('token-hash', $entity->getToken());
$this->assertEquals($queryCount, $this->getCurrentQueryCount());

$this->assertEquals('FabioBatSilva', $entity->getClient()->getName());
$this->assertEquals($queryCount, $this->getCurrentQueryCount());
```

With this, the test is green.

**Comment by bakura10** [comment]
Nope, you need to uncomment the ->clear() call. Actually the use case is the following: in first request, the entity is put in cache, in next request, obviously no object is in identity map.

**Comment by goetas** [comment]
@bakura10 I have edited my previous comment. Even for this version, tests are green.

It it ok?

**Comment by bakura10** [comment]
Yep, sounds nice :D.

**Comment by bakura10** [comment]
I just do not understand what happen under the hood. It seems that you are retrieving the association from the id, but in the case, the id does not exist because it has not been put into the cache.

**Comment by goetas** [comment]
@bakura10 I've edited the `buildCacheEntry` method, this means that I'm putting data into cache, later, `UoW::createEntity` will be able to create the right entity.

**Comment by bakura10** [comment]
But you are not supposed to put the id in cache if @Cache is not present in the association. For instance, if the client of the token is modified on the token, then you will retrieve an out-of-date association ID (or maybe Doctrine is smart enough to invalidate the cache whenever the entity is updated).

File: tests/Doctrine/Tests/Models/Cache/Action.php [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+     */
+    private $tokens;
+
+    public function __construct($name)
+    {
+        $this->name = $name;
+        $this->tokens = new ArrayCollection();
+    }
+
+    public function addToken(Token $token)
```
Comment (Ocramius) [inline tests/Doctrine/Tests/Models/Cache/Action.php]: Please remove all unnecessary API

File: tests/Doctrine/Tests/Models/Cache/Client.php [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+     * @Column(unique=true)
+     */
+    private $name;
+
+    public function __construct($name)
+    {
+        $this->name = $name;
+    }
+
+    public function getId()
```
Comment (Ocramius) [inline tests/Doctrine/Tests/Models/Cache/Client.php]: Same as above

File: tests/Doctrine/Tests/Models/Cache/Login.php [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+    public function __construct($name)
+    {
+        $this->name = $name;
+    }
+
+    /**
+     * @param Token $token
+     */
+    public function setToken(Token $token)
```
Comment (Ocramius) [inline tests/Doctrine/Tests/Models/Cache/Login.php]: Same as above

File: tests/Doctrine/Tests/Models/Cache/Action.php [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+class Action
+{
+    const CLASSNAME = __CLASS__;
+
+    /**
+     * @Id
+     * @GeneratedValue
+     * @Column(type="integer")
+     */
+    protected $id;
```
Comment (Ocramius) [inline tests/Doctrine/Tests/Models/Cache/Action.php]: Make properties `public` for testing

File: tests/Doctrine/Tests/Models/Cache/Token.php [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        $this->token     = $token;
+        $this->logins    = new ArrayCollection();
+        $this->client    = $client;
+        $this->expiresAt = new \DateTime(date('Y-m-d H:i:s', strtotime("+7 day")));
+    }
+
+    /**
+     * @return array
+     */
+    public function getLogins()
```
Comment (Ocramius) [inline tests/Doctrine/Tests/Models/Cache/Token.php]: Same here

File: tests/Doctrine/Tests/ORM/Functional/SecondLevelCacheOneToManyTest.php [hunk 8]
```diff
@@ -381,4 +383,34 @@ public function testCacheInitializeCollectionWithNewObjects()
...
+
+        $l1 = new Login('session1');
+        $l2 = new Login('session2');
+        $token  = new Token('token-hash');
+
+        $this->_em->persist($token);
+        $this->_em->flush();
+        $token->addLogin($l1);
+        $token->addLogin($l2);
+        $this->_em->flush();
```
Comment (Ocramius) [inline tests/Doctrine/Tests/ORM/Functional/SecondLevelCacheOneToManyTest.php]: Why is a second flush required here?

File: tests/Doctrine/Tests/ORM/Functional/SecondLevelCacheOneToManyTest.php [hunk 9]
```diff
@@ -381,4 +383,34 @@ public function testCacheInitializeCollectionWithNewObjects()
...
+        $this->assertTrue($this->cache->containsEntity(Token::CLASSNAME, $token->getToken()));
+
+        $entity = $this->_em->find(Token::CLASSNAME, $token->getToken());
+
+        $this->assertInstanceOf(Token::CLASSNAME, $entity);
+        $this->assertEquals('token-hash', $entity->getToken());
+        $this->assertEquals($queryCount, $this->getCurrentQueryCount());
+
+        $this->assertCount(2, $entity->getLogins());
+        $this->assertEquals($queryCount + 1 , $this->getCurrentQueryCount());
```
Comment (Ocramius) [inline tests/Doctrine/Tests/ORM/Functional/SecondLevelCacheOneToManyTest.php]: Remove that space before the comma

**Comment by goetas** [comment]
green :)

Even with foreign entities  as keys
/cc @guilhermeblanco

File: tests/Doctrine/Tests/Models/Cache/Action.php (hunk 3 above)
Comment (goetas) [inline tests/Doctrine/Tests/Models/Cache/Action.php]: The `addToken` method is used to create easily the link between "action" and "token" instances

**Comment by goetas** [comment]
@Ocramius I have updated the test suite

File: lib/Doctrine/ORM/Cache/DefaultEntityHydrator.php [hunk 10]
```diff
@@ -72,8 +81,31 @@ public function buildCacheEntry(ClassMetadata $metadata, EntityCacheKey $key, $e
...
+                $targetClassMetadata = $this->em->getClassMetadata($assoc['targetEntity']);
+                $associationIds = $this->identifierFlattener->flattenIdentifier($targetClassMetadata, $targetClassMetadata->getIdentifierValues($data[$name]));
+                unset($data[$name]);
+
+                foreach ($associationIds as $fieldName => $fieldValue) {
+
+                    if (isset($targetClassMetadata->associationMappings[$fieldName])){
+                        $targetAssoc = $targetClassMetadata->associationMappings[$fieldName];
+
+                        foreach($assoc['targetToSourceKeyColumns'] as $referencedColumn => $localColumn) {
```
Comment (goetas) [inline lib/Doctrine/ORM/Cache/DefaultEntityHydrator.php]: @guilhermeblanco To be honest I'm not sure that this approach will work with nested composite keys & associations

**Comment by Ocramius** [comment]
:+1:

**Pull request merged**